### PR TITLE
feat: Implement Logs Bridge API

### PR DESCRIPTION
**Pull request by kaylareopelle**
This PR provides an interface for:
* LoggerProvider
* Logger
* LogRecord

This API will be consumed by the forthcoming Logs SDK.

Follows specification version 1.23.0 

Relevant specs to this PR
* https://opentelemetry.io/docs/specs/otel/logs/bridge-api/ 
* https://opentelemetry.io/docs/specs/otel/logs/noop/
* https://opentelemetry.io/docs/specs/otel/logs/data-model/ (this is the source of the argument descriptions for Logger#create_log_record)

Closes: #1478
Closes: #1480
Closes: #1479
Closes: #1481

### CODE REVIEW

**Comment by kaylareopelle** [comment]
It looks like TruffleRuby may have an issue with SimpleCov's `:branch` coverage: https://github.com/open-telemetry/opentelemetry-ruby/actions/runs/5744430524/job/15570701339?pr=1509#step:7:150 

I'll try removing branch coverage and see if that helps.

File: logs_api/lib/opentelemetry/logs/logger_provider.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+      # @param schema_url [optional String] Schema URL recorded in the emitted telemetry.
+      # @param attributes [optional Hash] Specifies the instrumentation scope
+      #  attributes to associate with emitted telemetry.
+      #
+      # @return [OpenTelemetry::Logs::Logger]
+      #
+      # @api public
+      def logger(name: nil, version: nil, schema_url: nil, attributes: {})
+        @logger ||= OpenTelemetry::Logs::Logger.new
+      end
```
Comment (kaylareopelle) [inline logs_api/lib/opentelemetry/logs/logger_provider.rb:24]: I was chatting with @fallwith about this method. He asked something I hope folks more versed in the spec can help us answer:

> This method takes 4 named arguments that are all unused. Should the API version of this method not have them, have them with leading underscore-based names to signify non-use, or is the current state intentional for copying and pasting for non-no-op implementations?

Comment (alanwest) [reply logs_api/lib/opentelemetry/logs/logger_provider.rb:25]: Not a Ruby person, so I'm not really one to say what would seem right to a Ruby developer.. but I did note that Ruby's trace API does not use leading underscores:

https://github.com/open-telemetry/opentelemetry-ruby/blob/ccad50f9e5f04ba61fa991cabe795853ec87060b/api/lib/opentelemetry/trace/tracer_provider.rb#L17-L19

Also, this probably makes sense because this API will be consumed by instrumentation authors and whether or not the arguments will be used is unknown to them. Their use depends on an application author configuring the SDK for their application.

Comment (alanwest) [reply logs_api/lib/opentelemetry/logs/logger_provider.rb:25]: Also of note.. the trace API does not currently support `schema_url` or `attributes` for tracer provider. Might make sense to add this support across all signals at the same time. Does it make sense to remove these arguments for now from the logger provider?

File: logs_api/lib/opentelemetry/logs/logger_provider.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Logs
+    # No-op implementation of a logger provider
+    class LoggerProvider
+      # Returns a {Logger} instance.
+      #
```
Comment (alanwest) [inline logs_api/lib/opentelemetry/logs/logger_provider.rb:12]: The specification requires some kind of documentation that makes it clear that the log bridge API is not meant for direct consumption by application authors.

See: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md#artifact-naming

> The Logs Bridge API is not intended to be called by application developers directly, and SHOULD include documentation that discourages direct use.

Here's an example of how Java meets this requirement:

https://github.com/open-telemetry/opentelemetry-java/blob/90ab3665dc0fb56598fcf9b9c15941eedb3822f8/api/all/src/main/java/io/opentelemetry/api/logs/LoggerProvider.java#L14-L16

**Review by alanwest — COMMENTED** [review]
👋 Heyo, This looks pretty awesome @kaylareopelle! Excited to see some ❤️ for 🪵 in opentelemetry-ruby. Just a few thoughts/observations from a non-Ruby dev...

File: logs_api/lib/opentelemetry/logs/logger_provider.rb (hunk 1 above)
Comment (kaylareopelle) [reply logs_api/lib/opentelemetry/logs/logger_provider.rb:25]: > Also of note.. the trace API does not currently support schema_url or attributes for tracer provider. Might make sense to add this support across all signals at the same time. Does it make sense to remove these arguments for now from the logger provider?

Thanks for catching that! I think it'd make sense to drop them for now, but I'd like to hear more from the OTel Ruby team. 

cc @mwear, @robbkidd, @robertlaurin

File: logs_api/lib/opentelemetry/logs/logger_provider.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      # @param version [optional String] Instrumentation package version
+      # @param schema_url [optional String] Schema URL recorded in the emitted
+      #   telemetry.
+      # @param attributes [optional Hash] Specifies the instrumentation scope
+      #   attributes to associate with emitted telemetry.
+      #
+      # @return [OpenTelemetry::Logs::Logger]
+      #
+      # @api public
+      def logger(name: nil, version: nil, schema_url: nil, attributes: {})
```
Comment (kaylareopelle) [inline logs_api/lib/opentelemetry/logs/logger_provider.rb:23]: The methods I've added are using named arguments. It looks like the SDK and API use them more sparingly. Is there a preference for one approach over the other?

File: logs_api/lib/opentelemetry/logs/logger_provider.rb (hunk 2 above)
Comment (kaylareopelle) [reply logs_api/lib/opentelemetry/logs/logger_provider.rb:12]: Thank you! I'll update the documentation.

File: logs_api/lib/opentelemetry/logs/logger_provider.rb (hunk 1 above)
Comment (mwear) [reply logs_api/lib/opentelemetry/logs/logger_provider.rb:25]: I haven't had time to look at the PR as a whole, but I can answer this question with my thoughts (which may differ from others). I think we should stay consistent in our `*Provider` APIs. Given that we've gone with positional arguments for tracing, we should probably do the same for logs. I would also say that we should omit schema url for now, and add it to all signals, when the time is right.

File: logs_api/lib/opentelemetry/logs/logger.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      #   the severity as it is known at the source. Also known as log level.
+      # @param body [optional String, Numeric, Boolean, Array<String, Numeric,
+      #   Boolean>, Hash{String => String, Numeric, Boolean, Array<String,
+      #   Numeric, Boolean>}] A value containing the body of the log record.
+      # @param attributes [optional Hash{String => String, Numeric, Boolean,
+      #   Array<String, Numeric, Boolean>}] Additional information about the
+      #   event.
+      #
+      # @api public
+      def create_log_record(
```
Comment (mwear) [inline logs_api/lib/opentelemetry/logs/logger.rb:35]: I notice the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md#emit-a-logrecord) refers to "emitting" a log record.

I also noticed that [javascript](https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/api-logs/src/types/Logger.ts#L25) calls this method `emit` and that it doesn't return a value. It might be a good idea to see what implementations look like in other languages. 

My initial impression is that this method creates a LogRecord internally and passes it on, without returning it to the caller.

File: logs_api/lib/opentelemetry/logs/logger_provider.rb (hunk 3 above)
Comment (mwear) [inline logs_api/lib/opentelemetry/logs/logger_provider.rb:23]: The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/noop.md#logger-creation) allows us to return a single logger instance for the Noop implementation. It says:

> Since all Loggers will hold the same empty state, a LoggerProvider MAY return the same Logger instances to all creation requests.

We might be able to get away with defining a `NOOP_LOGGER = Logger.new` and always returning it. 

Not to complicate matters, but for tracing, it was expected that if an application started with the NoOp implementation, that tracers would become operational if an SDK was installed later. You can see the API installs a `ProxyTracerProvider` (which hands out `ProxyTracer`s). The delegate for the proxy tracers are initially API tracers, but they can be upgraded to SDK tracers.

I was trying to figure out if the same requirement exists for Logging, but I wasn't immediately able to track that down. We should find that out as that will impact what we want to do here.

If we do keep this as is, we can drop the `OpenTelemery::Logs::` namespacing.

**Review by mwear — COMMENTED** [review]
Ok, I got sniped into looking at this a little more. I have a few suggestions and / or questions.

File: logs_api/lib/opentelemetry/logs/logger.rb (hunk 4 above)
Comment (kaylareopelle) [reply logs_api/lib/opentelemetry/logs/logger.rb]: Thanks for bringing this up! 

[Python](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/_logs/_internal/__init__.py#L100C5-L102), [PHP](https://github.com/open-telemetry/opentelemetry-php/blob/main/src/API/Logs/LoggerInterface.php#L9) and [Java](https://github.com/open-telemetry/opentelemetry-java/blob/main/api/all/src/main/java/io/opentelemetry/api/logs/DefaultLogger.java#L79) all have similar implementations to JavaScript. 

I've updated the method to be named `#emit` and return nil in 12d39578d21852bf1926a4b5f954fbb52af2cf9a

File: logs_api/lib/opentelemetry/logs/logger_provider.rb (hunk 2 above)
Comment (kaylareopelle) [reply logs_api/lib/opentelemetry/logs/logger_provider.rb]: I've added a statement discouraging direct use by app developers in a006fba84e7c69d869403860e705b3184bdb9e17

Like Java, it's included in the Logger Provider and the Logger class definitions. Let me know if you think it should be included in all logger-related classes. I couldn't find similar notes in [PHP](https://github.com/open-telemetry/opentelemetry-php/blob/main/src/API/Logs/LoggerProviderInterface.php), [Python](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-api/src/opentelemetry/_logs/_internal/__init__.py), or [JavaScript](https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/api-logs/src/types/LoggerProvider.ts).

File: logs_api/lib/opentelemetry/logs/logger_provider.rb (hunk 3 above)
Comment (kaylareopelle) [reply logs_api/lib/opentelemetry/logs/logger_provider.rb]: Great idea. I'll do a bit of digging to find out whether the same requirement exists for Logs.

I've incorporated `NOOP_LOGGER` in dc966f0a153a386f91614742c3c6264422b9e6ca

File: logs_api/lib/opentelemetry/logs/logger_provider.rb (hunk 1 above)
Comment (kaylareopelle) [reply logs_api/lib/opentelemetry/logs/logger_provider.rb]: Thanks, @mwear! I've changed the code to be more consistent with the`TracerProvider` API, pulling out `schema_url` and `attributes` in dc966f0a153a386f91614742c3c6264422b9e6ca

File: logs_api/lib/opentelemetry/logs/logger_provider.rb (hunk 2 above)
Comment (kaylareopelle) [reply logs_api/lib/opentelemetry/logs/logger_provider.rb]: Maybe it should also be included in the definition of the `OpenTelemetry::Logs` module?

File: logs_api/lib/opentelemetry/logs/logger.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Logs
+    # No-op implementation of logger.
+    #
+    # The Logs Bridge API is provided for logging library authors to build
+    # log appenders/bridges. It should NOT be used directly by application
+    # developers.
```
[the remark this answers is not in the document]
Comment (mwear) [reply logs_api/lib/opentelemetry/logs/logger.rb:13]: I think we can probably just document this on `OpenTelemetry::Logs`.

File: logs_api/lib/opentelemetry/logs/logger_provider.rb [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+module OpenTelemetry
+  module Logs
+    # No-op implementation of a logger provider.
+    #
+    # The Logs Bridge API is provided for logging library authors to build
+    # log appenders/bridges. It should NOT be used directly by application
+    # developers.
+    class LoggerProvider
+      NOOP_LOGGER = OpenTelemetry::Logs::Logger.new
+
```
[the remark this answers is not in the document]
Comment (kaylareopelle) [reply logs_api/lib/opentelemetry/logs/logger_provider.rb:17]: Thank you!

**Review by mwear — APPROVED** [review]
LGTM. Thanks @kaylareopelle!

File: logs_api/lib/opentelemetry/logs/logger.rb [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      #   Numeric, Boolean>}] A value containing the body of the log record.
+      # @param attributes [optional Hash{String => String, Numeric, Boolean,
+      #   Array<String, Numeric, Boolean>}] Additional information about the
+      #   event.
+      #
+      # @api public
+      def emit(
+        timestamp: nil,
+        observed_timestamp: Process.clock_gettime(Process::CLOCK_REALTIME),
+        context: OpenTelemetry::Context.current,
```
Comment (dazuma) [inline logs_api/lib/opentelemetry/logs/logger.rb:39]: One quick drive-by comment: Perhaps we should default everything to `nil` so this doesn't actually make these calls if the SDK isn't present. If it's important to note the intended defaults, put it in the documentation.

**Comment by robertlaurin** [comment]
Approved with the expectation this comment is addressed before we merge https://github.com/open-telemetry/opentelemetry-ruby/pull/1509#discussion_r1294845162

File: logs_api/lib/opentelemetry/logs/logger.rb (hunk 7 above)
Comment (kaylareopelle) [reply logs_api/lib/opentelemetry/logs/logger.rb]: Great call-out! I've updated the code to include this recommendation.

**Comment by kaylareopelle** [comment]
@robertlaurin  - The comment has been addressed! [Here's the current file.](https://github.com/open-telemetry/opentelemetry-ruby/pull/1509/files#diff-7aa20ca852cf4c60647f9cecdb99ced26631799128923aba1ff0239908129c0d)
